rpcs3/rpcs3qt/elf_memory_dumping_dialog.cpp

```diff
@@ -0,0 +1,220 @@
+	{
+		bool ok = false;
+
+		// Parse expression (or at least used to, was nuked to remove the need for QtJsEngine)
+		const QString fixed_expression = QRegularExpression(QRegularExpression::anchoredPattern("a .*|^[A-Fa-f0-9]+$")).match(text).hasMatch() ? "0x" + text : text;
+		const u32 res = static_cast<u32>(fixed_expression.toULong(&ok, 16));
+
+		if (!ok)
+		{
+			errors << error_field;
+			return umax;
+		}
+
+		return res;
+	};
+
+	spu_memory_segment_dump_data data{};
+	data.segment_size = interpret(m_segment_size_input->text(), tr("Segment Size"));
+	data.src_addr = vm::get_super_ptr(interpret(m_ppu_address_input->text(), tr("PPU Address")));
+	data.ls_addr = interpret(m_ls_address_input->text(), tr("LS Address"));
+	data.flags = interpret(m_segment_flags_input->text(), tr("Segment Flags"));
+
+	if (!errors.isEmpty())
+	{
+		QMessageBox::warning(this, tr("Failed To Add Segment"), tr("Segment Parameters Are Incorrect: %1").arg(errors.join('\n')));
+		return;
+	}
+
+	if (data.segment_size % 4)
+	{
+		QMessageBox::warning(this, tr("Failed To Add Segment"), tr("SPU Segment Size Must Be 4 Bytes Aligned."));
+		return;
+	}
+
+	if (data.segment_size + data.ls_addr > SPU_LS_SIZE || data.segment_size == 0 || data.segment_size % 4)
+	{
+		QMessageBox::warning(this, tr("Failed To Add Segment"), tr("SPU Segment Range Is Invalid."));
+		return;
+	}
+
+	if (!vm::check_addr(vm::try_get_addr(data.src_addr).first, vm::page_readable, data.segment_size))
+	{
+		QMessageBox::warning(this, tr("Failed To Add Segment"), tr("PPU Address Range Is Not Accessible."));
+		return;
+	}
+
+	for (int i = 0; i < m_seg_list->count(); ++i)
+	{
+		const auto seg_stored = m_seg_list->item(i)->data(Qt::UserRole).value<spu_memory_segment_dump_data>();
+
+		const auto stored_max = seg_stored.src_addr + seg_stored.segment_size;
+		const auto data_max = data.src_addr + data.segment_size;
+
+		if (seg_stored.src_addr < data_max && data.src_addr < stored_max)
+		{
+			QMessageBox::warning(this, tr("Failed To Add Sehgment"), tr("SPU Segment Overlaps With Previous SPU Segment(s)\n"));
+			return;
+		}
+	}
+
+	auto item = new QListWidgetItem(tr("PPU Address: 0x%0, LS Address: 0x%1, Segment Size: 0x%2, Flags: 0x%3").arg(+vm::try_get_addr(data.src_addr).first, 2, 16).arg(data.ls_addr, 2, 16).arg(data.segment_size, 2, 16).arg(data.flags, 2, 16), m_seg_list);
+	item->setData(Qt::UserRole, QVariant::fromValue(data));
+	m_seg_list->setCurrentItem(item);
+}
+
+void elf_memory_dumping_dialog::remove_segment()
+{
+	const int row = m_seg_list->currentRow();
+	if (row >= 0)
+	{
+		QListWidgetItem* item = m_seg_list->takeItem(row);
+		delete item;
+	}
+}
+
+void elf_memory_dumping_dialog::save_to_file()
+{
+	std::vector<spu_memory_segment_dump_data> segs;
+	segs.reserve(m_seg_list->count());
+
+	for (int i = 0; i < m_seg_list->count(); ++i)
+	{
+		const auto seg_stored = m_seg_list->item(i)->data(Qt::UserRole).value<spu_memory_segment_dump_data>();
```

Comment:
missing sanity check for canConvert<spu_memory_segment_dump_data>

Reply:
Just curoius, what happens if it remains this way in case it can't? default initialized value?

Reply:
If the value cannot be converted, a default-constructed value will be returned.

Reply:
it's more like a check that the code works

Reply:
so ensure should be fine